### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by erlendcognite**

### CODE REVIEW

**Comment by cognite-cicd** [comment]
[Tests failed](https://jenkins.cognite.ai/job/github-builds/job/cognite-sdk-python/job/PR-108/4/testReport/).

#### test_tag_matching.test_object
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 21
  def test_object(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_json
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 26
  def test_json(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_pandas
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 30
  def test_pandas(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

... and 9 more test failures. Go to Jenkins for the full list.

**Comment by cognite-cicd** [comment]
[Tests failed](https://jenkins.cognite.ai/job/github-builds/job/cognite-sdk-python/job/PR-108/5/testReport/).

#### test_tag_matching.test_object
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 21
  def test_object(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_json
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 26
  def test_json(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_pandas
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 30
  def test_pandas(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

... and 9 more test failures. Go to Jenkins for the full list.

**Comment by cognite-cicd** [comment]
[Tests failed](https://jenkins.cognite.ai/job/github-builds/job/cognite-sdk-python/job/PR-108/6/testReport/).

#### test_tag_matching.test_object
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 21
  def test_object(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_json
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 26
  def test_json(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_pandas
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 30
  def test_pandas(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

... and 9 more test failures. Go to Jenkins for the full list.

**Comment by cognite-cicd** [comment]
[Tests failed](https://jenkins.cognite.ai/job/github-builds/job/cognite-sdk-python/job/PR-108/8/testReport/).

#### test_tag_matching.test_object
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 21
  def test_object(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_json
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 26
  def test_json(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

#### test_tag_matching.test_pandas
```
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 30
  def test_pandas(tagmatching_result):
file /home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py, line 9
  @patch('requests.post')
  @pytest.fixture(scope='module')
  def tagmatching_result(mock_post):
E       fixture 'mock_post' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, configure_test_session, cov, doctest_namespace, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, tagmatching_result, tmpdir, tmpdir_factory, unset_config_variables
>       use 'pytest --fixtures [testpath]' for help on them.

/home/jenkins/workspace/s_cognite-sdk-python_PR-108-ZOFF24QXX7SGAHE7IUMSMHATVGGA32EZPQGQNYI6QNQMJMLNEUMQ/tests/v04/test_tag_matching.py:9
```

... and 9 more test failures. Go to Jenkins for the full list.

File: cognite/auxiliary/_hosting/model/model/model.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Model:
+    def __init__(self, model):
+        self._model = model
+
+    @staticmethod
+    def train(file_io, **kwargs):
+        """The method to train your model.
+        Should produce at least a serialized model. Can also produce
```
Comment (NilsBarlaug-cognite) [inline cognite/auxiliary/_hosting/model/model/model.py]: Can also produce ... ?

File: cognite/cli/cli.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class CogniteCLI:
+    def __init__(self):
+        """cognite <service> [<args>]
+
+The available Cognite services are:
+    models      Model hosting
+"""
+        parser = argparse.ArgumentParser(description="Command line interface for CDP.", usage=self.__init__.__doc__)
+        parser.add_argument("service", help="Service to run", choices=["models"])
```
Comment (NilsBarlaug-cognite) [inline cognite/cli/cli.py]: Maybe you can use sub-commands? https://docs.python.org/3/library/argparse.html#sub-commands

File: cognite/cli/cli.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        # parse_args defaults to [1:] for args, but you need to
+        # exclude the rest of the args too, or validation will fail
+        args = parser.parse_args(sys.argv[1:2])
```
Comment (NilsBarlaug-cognite) [inline cognite/cli/cli.py]: If you use sub-commands this can be solved with set_defaults()

File: cognite/cli/cli.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        if not hasattr(self, args.service.replace("-", "_")):
+            print("Unrecognized service")
+            parser.print_help()
+            exit(1)
+        # use dispatch pattern to invoke method with same name
+        getattr(self, args.service)(sys.argv[2:])
```
Comment (NilsBarlaug-cognite) [inline cognite/cli/cli.py]: If you use sub-commands this can be solved with set_defaults()

File: cognite/cli/cli.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    def models(self, args):
+        """cognite models <command> [<args>]
+
+The available commands are:
+    source      Create a new model source package in the current directory.
+    go          Run tests, upload a source package and create a model with the same name.
```
Comment (NilsBarlaug-cognite) [inline cognite/cli/cli.py]: Maybe we come up with a more descriptive word?

Comment (erlendcognite) [reply cognite/cli/cli.py]: @NilsBarlaug-cognite

**Pull request merged**